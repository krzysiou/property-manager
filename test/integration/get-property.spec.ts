import assert from 'assert';

import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import type { Property } from '../../src/app/schemas/types.codegen.js';

import { pupulateDatabase } from '../utils/database-utility/populate-database.js';
import { pruneDatabase } from '../utils/database-utility/prune-database.js';
import { assumeType } from '../utils/assume-type.js';
import { getTestDatabase, getTestServer } from './index.js';

describe('integration/get-property', () => {
  const testServer = getTestServer();
  const database = getTestDatabase();

  let mockIds: string[];

  beforeEach(async () => {
    mockIds = await pupulateDatabase(database);
  });

  afterEach(async () => {
    await pruneDatabase(mockIds, database);
  });

  it('returns property data if property exists', async () => {
    const response = await testServer.executeOperation({
      query: `
      query GetProperty($getPropertyId: ID!) { 
        getProperty(id: $getPropertyId) { 
          id
        } 
      }`,
      variables: { getPropertyId: mockIds[0] },
    });

    assert(response.body.kind === 'single');

    const data = assumeType<Property>(
      response.body.singleResult.data?.getProperty
    );

    expect(data.id).toEqual(mockIds[0]);
  });

  it('returns null data if property does not exist', async () => {
    const response = await testServer.executeOperation({
      query: `
      query GetProperty($getPropertyId: ID!) { 
        getProperty(id: $getPropertyId) { 
          id
        } 
      }`,
      variables: { getPropertyId: 'invalid-id' },
    });

    assert(response.body.kind === 'single');

    const data = assumeType<Property>(
      response.body.singleResult.data?.getProperty
    );

    expect(data).toBeNull();
  });

  it('returns property data if property exists', async () => {
    const response = await testServer.executeOperation({
      query: `
      query GetProperty($getPropertyId: ID!) { 
        getProperty(id: $getPropertyId) { 
          id
          city
          street
          state
          zipCode
          lon
          lat
          creationDate
          weatherData {
            observationTime
            temperature
            weatherCode
            windSpeed
            windDegree
            windDir
            pressure
            precip
            humidity
            cloudcover
            feelslike
            uvIndex
            visibility
          } 
        } 
      }`,
      variables: { getPropertyId: mockIds[0] },
    });

    assert(response.body.kind === 'single');

    const data = assumeType<Property>(
      response.body.singleResult.data?.getProperty
    );

    expect(data).toMatchSnapshot({
      id: expect.any(String),
    });
  });
});
