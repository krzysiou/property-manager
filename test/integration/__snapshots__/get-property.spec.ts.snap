// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration/get-property > returns property data if property exists 1`] = `
{
  "city": "New York",
  "creationDate": "2025-01-18T19:26:20.965Z",
  "id": Any<String>,
  "lat": "5",
  "lon": "-5",
  "state": "AK",
  "street": "Angel st.",
  "weatherData": {
    "cloudcover": 20,
    "feelslike": 23,
    "humidity": 65,
    "observationTime": "07:24 PM",
    "precip": 0,
    "pressure": 1013,
    "temperature": 22,
    "uvIndex": 5,
    "visibility": 10,
    "weatherCode": 1,
    "windDegree": 270,
    "windDir": "W",
    "windSpeed": 15,
  },
  "zipCode": "12345",
}
`;
